File: java/src/test/java/algorithms/arrays/MaximumAverageSubarrayTest.java
package algorithms.arrays;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class MaximumAverageSubarrayTest {
    
    private final MaximumAverageSubarray solution = new MaximumAverageSubarray();
    
    @Test
    void testFindMaxAverage_basicCase() {
        int[] nums = {1, 12, -5, -6, 50, 3};
        int k = 4;
        
        double result = solution.findMaxAverage(nums, k);
        
        assertEquals(12.75, result, 0.00001);
    }
    
    @Test
    void testFindMaxAverage_singleElement() {
        int[] nums = {5};
        int k = 1;
        
        double result = solution.findMaxAverage(nums, k);
        
        assertEquals(5.0, result, 0.00001);
    }
    
    @Test
    void testFindMaxAverage_entireArray() {
        int[] nums = {1, 2, 3, 4};
        int k = 4;
        
        double result = solution.findMaxAverage(nums, k);
        
        assertEquals(2.5, result, 0.00001);
    }
    
    @Test
    void testFindMaxAverage_negativeNumbers() {
        int[] nums = {-1, -2, -3, -4, -5};
        int k = 2;
        
        double result = solution.findMaxAverage(nums, k);
        
        assertEquals(-1.5, result, 0.00001);
    }
    
    @Test
    void testFindMaxAverage_consecutivePositives() {
        int[] nums = {0, 1, 1, 3, 3};
        int k = 4;
        
        double result = solution.findMaxAverage(nums, k);
        
        assertEquals(2.0, result, 0.00001);
    }
}

File: elixir/test/algorithms/arrays/maximum_average_subarray_test.exs
defmodule Algorithms.Arrays.MaximumAverageSubarrayTest do
  use ExUnit.Case
  alias Algorithms.Arrays.MaximumAverageSubarray

  test "basic case" do
    result = MaximumAverageSubarray.find_max_average([1, 12, -5, -6, 50, 3], 4)
    assert_in_delta result, 12.75, 0.00001
  end

  test "single element" do
    result = MaximumAverageSubarray.find_max_average([5], 1)
    assert_in_delta result, 5.0, 0.00001
  end

  test "entire array" do
    result = MaximumAverageSubarray.find_max_average([1, 2, 3, 4], 4)
    assert_in_delta result, 2.5, 0.00001
  end

  test "negative numbers" do
    result = MaximumAverageSubarray.find_max_average([-1, -2, -3, -4, -5], 2)
    assert_in_delta result, -1.5, 0.00001
  end

  test "consecutive positives" do
    result = MaximumAverageSubarray.find_max_average([0, 1, 1, 3, 3], 4)
    assert_in_delta result, 2.0, 0.00001
  end

  test "prefix sum implementation works" do
    result = MaximumAverageSubarray.find_max_average_prefix([1, 12, -5, -6, 50, 3], 4)
    assert_in_delta result, 12.75, 0.00001
  end
end
